server/server/src/billing/billing_service.rs

```diff
@@ -0,0 +1,396 @@
+}
+
+fn get_active_subscription_index<U: Debug>(
+    subscriptions: &[StripeSubscriptionInfo],
+) -> Result<usize, ServerError<U>> {
+    let active_pos = subscriptions
+        .iter()
+        .position(|info| info.is_active)
+        .ok_or_else(|| internal!("Redis says there is no active subscription despite the user having non free data cap: {:?}", subscriptions))?;
+
+    Ok(active_pos)
+}
+
+const MAXIMUM_PAYMENT_FLOW_LOCK_MILLIS: i64 = 30000;
+
+async fn lock_payment_workflow(
+    public_key: &PublicKey, con: &mut deadpool_redis::Connection,
+) -> Result<StripeUserInfo, ServerError<SwitchAccountTierError>> {
+    let mut user_info = StripeUserInfo::default();
+
+    let tx_result = tx!(con, pipe_name, &[stripe_user_info(public_key)], {
+        user_info = con
+            .maybe_json_get(stripe_user_info(public_key))
+            .await?
+            .unwrap_or_default();
+
+        if get_time().0 - user_info.last_in_payment_flow < MAXIMUM_PAYMENT_FLOW_LOCK_MILLIS {
+            return Err(Abort(ClientError(SwitchAccountTierError::AlreadyInBillingWorkflow)));
+        }
+
+        user_info.last_in_payment_flow = get_time().0;
+
+        pipe_name.json_set(stripe_user_info(public_key), &user_info)?;
+
+        Ok(&mut pipe_name)
+    });
+    return_if_error!(tx_result);
+
+    Ok(user_info)
+}
+
+async fn create_subscription(
+    server_state: &ServerState, con: &mut deadpool_redis::Connection, public_key: &PublicKey,
+    payment_method: &PaymentMethod, user_info: &mut StripeUserInfo,
+) -> Result<u64, ServerError<SwitchAccountTierError>> {
+    let (customer_id, payment_method_id) = match payment_method {
+        PaymentMethod::NewCard { number, exp_year, exp_month, cvc } => {
+            let payment_method_resp = stripe_client::create_payment_method(
+                &server_state.stripe_client,
+                number,
+                *exp_month,
+                *exp_year,
+                cvc,
+            )
+            .await?;
+
+            let customer_id = match &user_info.customer_id {
+                None => {
+                    let customer_resp = stripe_client::create_customer(
+                        &server_state.stripe_client,
+                        payment_method_resp.id(),
+                    )
+                    .await?;
+                    let customer_id = customer_resp.id.to_string();
+
+                    con.json_set(public_key_from_stripe_customer_id(&customer_id), public_key)
+                        .await?;
+
+                    user_info.customer_id = Some(customer_id);
+                    customer_resp.id
+                }
+                Some(customer_id) => stripe::CustomerId::from_str(customer_id)?,
+            };
+
+            if let Some(info) = user_info
+                .payment_methods
+                .iter()
+                .max_by_key(|info| info.created_at)
+            {
+                stripe_client::detach_payment_method_from_customer(
```

Comment:
I am detaching payment methods because there is a [max amount that can be added to a customer](https://stripe.com/docs/error-codes#customer-max-payment-methods).